### PR TITLE
Initialize tank temps

### PR DESCRIPTION
**Pull request by spahrenk**
## Description
- Added function setTankLayerTemperatures to enable setting Ts to each tank node.
- This addresses an issue that was discussed in a meeting with Chip about matching expt and sim data.
- The function accepts a vector<double> representing an array of temperatures. These are distributed across all nodes using uniform scaling. The existing setTankToTemperature function was modified to use this new function, to avoid redundancy.
- A general `resample` function was added, which calls the function `sample` to transform each element. Variants `resampleIntensive` and `resampleExtensive` address normalization.
-- The input vector is passed as a (non-const) parameter, rather than by reference, to allow unit changes, if needed. 
-- Binning is applied when upsampling. This makes the algorithm more difficult to analyze, and may impose a performance hit for simple cases, but could improve speed for large, mismatched vector sizes.
- Comments were added.
- Additional tests were incorporated into testMaxSetpoint.cc. 
- Checks are made for an empty input vector or a null node array pointer.

## Author Progress Checklist:

- [X] Open draft pull request
    - [X] Make title clearly understandable in a standalone change log context
    - [X] Assign yourself the issue
    - [X] Add at least one label (enhancement, bug, or maintenance)
    - [ ] Link the issue(s) addressed by this PR (under "Development" in the sidebar menu) 
- [X] Make code changes (if you haven't already)
- [X] Self-review of code
    - [X] My code follows the style guidelines of this project
    - [X] I have added comments to my code, particularly in hard-to-understand areas
    - [X] I have only committed the necessary changes for this fix or feature
    - [X] I have made corresponding changes to the documentation
    - [X] My changes generate no new warnings
    - [X] I have ensured that my fix is effective or that my feature works as intended by:
        - [ ] exercising the code changes in the test framework, and/or
        - [X] manually verifying the changes (as explained in the the pull request description above)
    - [X] My changes pass all local tests
    - [X] My changes successfully passes CI checks
    - [X] Add any unit test for proof and documentation.
    - [X] Merge in main branch and address resulting conflicts and/or test failures.
- [X] Move pull request out of draft mode and assign reviewers
- [] Iterate with reviewers until all changes are approved
    - [ ] Make changes in response to reviewer comments
    - [ ] Merge in main branch and address resulting conflicts and/or test failures.
    - [ ] Re-request review in GitHub

## Reviewer Checklist:

 - [ ] Read the pull request description
 - [ ] Perform a code review on GitHub
 - [ ] Confirm all CI checks pass and there are no build warnings
 - [ ] Pull, build, and run automated tests locally
 - [ ] Perform manual tests of the fix or feature locally
 - [ ] Add any review comments, if applicable
 - [ ] Submit review in GitHub as either
     - [ ] Request changes, or
     - [ ] Approve
 - [ ] Iterate with author until all changes are approved

### CODE REVIEW

**Comment by spahrenk** [comment]
Recent changes more carefully distribute the temperature inputs over the nodes when using setTankLayerTemperatures. It should only make a difference if the numbers of setpoint values and nodes don't divide evenly.

**Review by nealkruis — CHANGES_REQUESTED** [review]
@spahrenk this works as advertised, but the logic is very hard to follow. Can you take another pass with longer self-documenting variable names, and comments where it is not obvious from the names what is happening?

Partly, I'm wondering if there might be a cleaner (and more straightforward way to perform the same functionality). I wonder, too, if there are maybe some trivial cases that we can handle more simply and with less computation (e.g., number of temperatures = number of nodes).

Finally, it would be good to demonstrate this is working with some fairly basic test cases (e.g., 1 temperature with many nodes, many temperatures with 1 node) to make sure we're covering the possibilities. We really should set up unit tests in this repository, but that might be a task for a future branch.

**Comment by spahrenk** [comment]
Yes, I agree with all of that. It is difficult to follow the logic and I expect that there are cases that can be handled more efficiently. I need to use a similar approach for the variable # of nodes issue, which I am already tackling on another branch. Then I will come and improve this.

**Comment by nealkruis** [comment]
Thanks, @spahrenk! This change is a bit higher priority, since we'd like to get it into a release of CSE soon. If at all possible, it would be great if you can address this one first.

**Comment by spahrenk** [comment]
I have attempted to improve this algorithm by adding comments. One special case is handled uniquely.

**Comment by spahrenk** [comment]
I simplified the resampling function to enable sampling a single point, which is needed in some cases. It just applies the point version to each vector value. But then I generalized again to offer some possible performance gains for large, mismatched datasets, which probably reduces the performance slightly for simpler datasets. Tried to improve variable names and comment formatting.

File: test/testMaxSetpoint.cc [hunk 1]
```diff
@@ -203,4 +207,80 @@ void testSetpointFixed() {
...
-}
\ No newline at end of file
+}
+
+void testResampling() {
+
+// test
+	std::vector<double> oldHeat{20., 40., 60., 40., 20.};
+	std::vector<double> newHeat(10, 0.);
+	resampleExtensive(newHeat, oldHeat);
```
Comment (tanaya-mankad) [inline test/testMaxSetpoint.cc]: The signature of resampleExtensive is `int resample(std::vector<double> &origValues,const std::vector<double> &newValues)` but this call seems to be using (new, old). Is this a naming issue or are the arguments reversed?

File: src/HPWH.cc [hunk 2]
```diff
@@ -71,10 +71,123 @@ const std::string HPWH::version_maint = HPWHVRSN_META;
 // setpoint-below-water-temp issues
 //   1-22-2017
 
+//-----------------------------------------------------------------------------
```
Comment (tanaya-mankad) [inline src/HPWH.cc]: I really like the function level comments, but let's stick with [our standard Doxygen](https://github.com/bigladder/atheneum/blob/main/doc/guidelines.md#comments) style. (The comment style itself is up for discussion, I've used the C++ comment style with "@" tags for years so that's what I put in the BL guidelines.) See [doxygen manual](https://www.doxygen.nl/manual/commands.html) for the insanely long list of tags that can be used.

File: src/HPWH.cc [hunk 3]
```diff
@@ -71,10 +71,123 @@ const std::string HPWH::version_maint = HPWHVRSN_META;
...
+	double totValueWeight = 0.;
+	double totWeight = 0.;
+	for(std::size_t i = i0; fracNext < fracEnd; ++i)
+	{
+		fracNext = static_cast<double>(i + 1) / nNodes;
+		if(fracNext > fracEnd)
+		{
+			fracNext = fracEnd;
+		}
+		double weight = fracNext - fracPrev;
```
Comment (tanaya-mankad) [inline src/HPWH.cc]: I might be misinterpreting this, but it seems that for every iteration except the first one, the weight is just the inverse of nNodes - so, a constant?

File: src/HPWH.cc [hunk 4]
```diff
@@ -71,10 +71,123 @@ const std::string HPWH::version_maint = HPWHVRSN_META;
...
+//	Returns: double: resampled value; 0 if undefined.
+//-----------------------------------------------------------------------------
+double sample(const std::vector<double> &values,double fracBegin,double fracEnd)
+{
+	if(fracBegin > fracEnd)std::swap(fracBegin,fracEnd);
+	if(fracBegin < 0.) fracBegin = 0.;
+	if(fracEnd >1.) fracEnd = 1.;
+
+	double nNodes = static_cast<double>(values.size());
+	auto i0 = static_cast<std::size_t>(fracBegin * nNodes);
```
Comment (tanaya-mankad) [inline src/HPWH.cc]: Could we do 'index' instead of 'i'?

File: src/HPWH.cc [hunk 5]
```diff
@@ -71,10 +71,123 @@ const std::string HPWH::version_maint = HPWHVRSN_META;
...
+//	Note(s): Bounding fractions are clipped or swapped, if needed.
+//	Returns: double: resampled value; 0 if undefined.
+//-----------------------------------------------------------------------------
+double sample(const std::vector<double> &values,double fracBegin,double fracEnd)
+{
+	if(fracBegin > fracEnd)std::swap(fracBegin,fracEnd);
+	if(fracBegin < 0.) fracBegin = 0.;
+	if(fracEnd >1.) fracEnd = 1.;
+
+	double nNodes = static_cast<double>(values.size());
```
Comment (tanaya-mankad) [inline src/HPWH.cc]: nNodes doesn't have to be a double type, the calculations that use it will automatically promote it if the other operand is double.

File: src/HPWH.cc [hunk 6]
```diff
@@ -71,10 +71,123 @@ const std::string HPWH::version_maint = HPWHVRSN_META;
...
+			fracNext = fracEnd;
+		}
+		double weight = fracNext - fracPrev;
+		totValueWeight += weight * values[i];
+		totWeight += weight;
+		fracPrev = fracNext;
+	}
+	double res = 0.;
+	if(totWeight > 0.) res = totValueWeight / totWeight;
+	return res;
```
Comment (tanaya-mankad) [inline src/HPWH.cc]: If a function returns a value and getting that value is the whole point of calling it, its name should reflect that, so maybe something like `get_weighted_resampled_value` (or better). Also `res` could be more descriptive.

File: src/HPWH.cc [hunk 7]
```diff
@@ -71,10 +71,123 @@ const std::string HPWH::version_maint = HPWHVRSN_META;
...
+//-----------------------------------------------------------------------------
+//    Description:
+//        Replaces the values in a std::vector by resampling another std::vector of
+//        arbitrary size.
+//    Parameter(s):
+//        std::vector<double> &origValues: contains values to be replaced
+//        const std::vector<double> &newValues: contains values to replace with
+//    Returns: int: 0: success; -1: failure.
+//-----------------------------------------------------------------------------
+int resample(std::vector<double> &origValues,const std::vector<double> &newValues)
```
Comment (tanaya-mankad) [inline src/HPWH.cc]: The return value here seems to indicate only success or failure, so I'd choose a bool.

File: src/HPWH.cc [hunk 8]
```diff
@@ -71,10 +71,123 @@ const std::string HPWH::version_maint = HPWHVRSN_META;
...
+    if(newValues.empty()) return -1;
+    double origSize = static_cast<double>(origValues.size());
+    double newSize = static_cast<double>(newValues.size());
+    double sizeRatio = newSize / origSize;
+    auto binSize = static_cast<std::size_t>(1. / sizeRatio);
+    double fracBegin = 0., fracEnd;
+    std::size_t i = 0;
+    while(i < origSize)
+    {
+        auto xi = static_cast<double>(i);
```
Comment (tanaya-mankad) [inline src/HPWH.cc]: Could we flesh out the variable names like i, j, xi, etc?

File: test/testMaxSetpoint.cc (hunk 1 above)
Comment (spahrenk) [reply test/testMaxSetpoint.cc]: Altered naming for clarity.

File: src/HPWH.cc (hunk 4 above)
Comment (spahrenk) [reply src/HPWH.cc]: OK. That starts to look too verbose for my taste, though.

File: src/HPWH.cc (hunk 5 above)
Comment (spahrenk) [reply src/HPWH.cc]: On the other hand, there is minimal penalty for casting it during the initial assignment, so I prefer to leave it as is.

File: src/HPWH.cc (hunk 7 above)
Comment (spahrenk) [reply src/HPWH.cc]: I was intending to be consistent with the existing program, but OK. Could be some other return codes in future versions, however.

File: src/HPWH.cc (hunk 3 above)
Comment (spahrenk) [reply src/HPWH.cc]: It can change on each iteration, in some cases.

File: src/HPWH.cc (hunk 7 above)
Comment (tanaya-mankad) [reply src/HPWH.cc]: When we need return codes, we can change it back. In the meantime we shouldn't need to reflect C-style practices.

**Comment by spahrenk** [comment]
Needed to be merged prior to any-number-of-nodes-basic.
